File: iterator-pattern/numbers-collection/numbers-collection.ts
class NumberCollection {
  private numbers: number[];

  constructor(nums: number[]) {
    this.numbers = nums;
  }

  [Symbol.iterator](): Iterator<number> {
    return new NumberIterator(this.numbers);
  }
}

class NumberIterator implements Iterator<number> {
  private numbers: number[];
  private index: number;

  constructor(nums: number[]) {
    this.numbers = nums;
    this.index = 0;
  }

  next(): IteratorResult<number> {
    if (this.index < this.numbers.length) {
      return { value: this.numbers[this.index++], done: false };
    } else {
      return { value: undefined, done: true };
    }
  }

  [Symbol.iterator](): Iterator<number> {
    return this;
  }
}

// Usage
const col = new NumberCollection([8, 12, 16, 18, 20]);

for (const num of col) {
  console.log(num);
}

File: iterator-pattern/numbers-collection/numbers-collection.py
from typing import List


class NumberCollection:
    """A simple collection of numbers."""

    def __init__(self, numbers: List[float]):
        self.numbers = numbers

    def __iter__(self):
        """Return the iterator object."""
        return NumberIterator(self.numbers)


class NumberIterator:
    """Iterator for the NumberCollection."""

    def __init__(self, numbers: List[float]):
        self.numbers = numbers
        self.index = 0

    def __iter__(self):
        return self

    def __next__(self):
        """Returns the next element or raises StopIteration if no elements left."""
        if self.index < len(self.numbers):
            result = self.numbers[self.index]
            self.index += 1
            return result
        else:
            raise StopIteration


# Dryrun
collection = NumberCollection([10, 20, 30])

for number in collection:
    print(number)
